### ISSUE TITLE
[meta]: 0.9 new UI changes review

### ISSUE THREAD

**Issue by GeopJr** [GeopJr/Tuba#1155]
@bertob @bragefuglseth 

I was planning on making this issue when I finished the other WIP features but since I've got both of your attention right now, I might as well start:

- #1099 Audio visualizer

https://github.com/user-attachments/assets/05aa7baa-4777-4870-960d-24f9ce0f971a

Follows accent, if there's a cover, it uses its average color instead

- #1101 Focus picker

https://github.com/user-attachments/assets/61d1f998-6f41-438c-bd2b-7ca53f7385f8

![image](https://github.com/user-attachments/assets/a998eda4-113e-4fdd-bdc1-bcde43d79399)

(focus picker icon top right)

- #1106 Mute dialog with options

![image](https://github.com/user-attachments/assets/92881f20-f2b0-404c-a7c5-23a3c498cbb5)

- #1112 Verified author(s)

![image](https://github.com/user-attachments/assets/d63d24fe-60bf-4fe6-8fab-6bd087fb02df)

flat button, on click opens profile or author link if available. The API *can* return multiple authors even though it's currently not done. In that case it will show "Show all %d authors" and open a popover with all of them.

- #1128 thumbnail in alt text dialog composer

![image](https://github.com/user-attachments/assets/9880596a-fc3b-4ef0-b073-1bab12df2a41)

- #1125 scale revealer stop animation early and ease in-out not just out

Difficult to explain, but #1122 - the way scale revealer works is by crossfading between the attachment and the mediaviewer. That has the effect of the attachment stretching to fit the screen around the mid point of the animation. That PR makes the animation cross fade early and makes it ease-in-out so it hits more points while crossfading, otherwise it might go from 0% to 29% and crossfade only for one frame

https://github.com/user-attachments/assets/2a36b76e-6fc8-4995-a5ca-1b367f5ce072

- #1147 auto-fill in advanced search dialog for user

[Screencast from 2024-10-04 21-09-32.webm](https://github.com/user-attachments/assets/b6796840-66ef-4936-9d16-ba2ac60f7628)

basically a button to search and return the first account in the advanced search dialog

- #1160 poll show results button and refresh results button

![Screenshot From 2024-10-13 19-21-56](https://github.com/user-attachments/assets/4dc2c0f1-96d3-4773-9baa-fd6395463672)
![Screenshot From 2024-10-13 19-22-03](https://github.com/user-attachments/assets/f0aa8e2c-33fa-48cd-bdfb-5ec2cc07fe88)

- #1161 poll progress animation

https://github.com/user-attachments/assets/38973349-27e5-44b7-a64e-efccb32ec513

- #1198  Wrapped

https://github.com/user-attachments/assets/8c9b33ca-be47-4ba1-a0bd-d8d0bb3921e6

Yearly recap of the user's mastodon activity. This is generated by the server and I have some more info on the linked PR. I went with an animated border on the avatar to make it a bit more fun (?). The 'Most Used App' card will only look like in the recording if it's 'Tuba', otherwise it will look like the rest and not have an animated background or colors.

- #1188 AdwToggleGroup-like profile filter switcher

https://github.com/user-attachments/assets/50406838-ad90-4139-8587-4930e6031859

AdwToggleGroup will be available in the next libadwaita so in the meantime I ported it from Elastic

- #1209 scheduled posts

Scheduled Posts view:

![Screenshot From 2024-11-22 00-27-24](https://github.com/user-attachments/assets/b8096953-9f41-43cb-9d57-4a6dac79b99e)

You can re-schedule them or delete them

Scheduling from the composer:

https://github.com/user-attachments/assets/7645729a-aa53-47fa-935a-388647367368

(change since recording: the re-schedule dialog shows 'Reschedule' as the headerbar button label)

Drafting will also be added so it wont be the only option in the composer dropdown

- #1216 Moved accounts notice

https://github.com/user-attachments/assets/dc817d6d-ea03-4fb9-aca1-aa52600552d5

- #1213 draft posts were added

They act similarly to schedule posts

---

(for a recap of everything to 0.9, https://github.com/GeopJr/Tuba/compare/v0.8.3...main)

**Comment by GeopJr** [comment]
Bump, reached feature freeze

**Comment by bertob** [comment]
Sorry about the delay, had a quick look at the latest main!

### Audio player

> Follows accent, if there's a cover, it uses its average color instead

If we can do average color, maybe it'd be better to do that for the user avatar as well?

### Focal Point Setting
I was wondering if the numeric inputs are needed? My feeling is this is usually not something that needs to be particularly precise, and it makes the dialog look a bit clunky.

![image](https://github.com/user-attachments/assets/e88102f6-e111-4c10-9cd0-6b59a100b633)

Thanks for working on this btw, I personally use this feature a lot!

### Composer Dialog Layouts
The save button in the headerbar right next to close feels a bit off. I'd either use the classic layout with "Cancel" on the left, or move "Save" out of the headerbar (e.g. to the bottom).

I've noticed that the titles for some of these dialogs are longer than they need to be (e.g. could be just "Alternative Text" and "Set Focal Point"). They should also use Title Case.

### Verified Authors
From that screenshot it looks a bit odd, but hard to judge it without more context. Do you have a link to a post with this? Also, is there any relevant art for this from other clients?

### Alt Text Composer Thumbnail
It'd be good to have a little bit of padding around the thumbnail, and a slight border radius. I'd also drop the different background color on the textarea.

![image](https://github.com/user-attachments/assets/1dae8c76-b3f0-4b55-b5c1-eb3aedad03a7)

### User Search/Autocomplete
This looks fun, but not very discoverable. Maybe instead of the button there could be an autocomplete popover with a list of search results that updates as you type the search query?

### Wrapped
Niiice, very cute feature :heart: 

The Default/Accent/etc. popovers are a bit confusing, both because it's not clear what they affect, and also because they're duplicated in two places. IMO it'd be clearer if there was a single "Style" chooser popover for this (ideally with some kind of inline preview), which would then be used if you download or share it. Maybe the background could be applied to this dialog as well, for an even clearer preview?

The share image needs a bit of padding around the widgets, e.g. 24px or so:

![2024-wrapped](https://github.com/user-attachments/assets/b835f5a0-3c81-4b6f-a7e1-298fd605b984)

The notification I got has "(null)" in it:

![Screenshot From 2024-12-02 23-39-42](https://github.com/user-attachments/assets/e645fa3d-dc9d-41d6-8bf4-9ee24435c094)

For my account the "most interactions" thing doesn't seem to work, I guess there should be a better empty state (e.g. "No Interactions" with an icon or something), or the card should be emtpy/hidden. 

![image](https://github.com/user-attachments/assets/2b1e1f8d-4394-4639-bb9d-834f2c43d3a0)

### Scheduling
The headerbar is a bit awkward here because of the long strings. Are you sure the button needs to say "Schedule and Publish"? My feeling is just "Schedule" would be enough, but maybe it could also say "Publish Later"? As for the title, I think it could just be "Schedule Post", no?

![image](https://github.com/user-attachments/assets/49786ee9-315f-4e78-a359-71c662be4bcf)

I'd also go for a more standard time chooser (like e.g. the one in Railway), the multi-row layout is a bit hard to parse.

![image](https://github.com/user-attachments/assets/99ca4a8d-be8d-4a93-9ff0-9a241c8581b2)

The row layout for scheduled posts would look better with a divider line between the header and the post itself. I'd also drop the bold and the colon in the header, and just go for a normal label like "Scheduled for March 20, 2025 ...".

![image](https://github.com/user-attachments/assets/381a6c95-1f5b-49dc-8bff-bc8b72131470)


Everything else looks good to me. Thanks for all your work on this, this app is really coming along nicely!

**Comment by GeopJr** [comment]
> Sorry about the delay, had a quick look at the latest main!

No problem, thanks for the review!

> If we can do average color, maybe it'd be better to do that for the user avatar as well?

Currently, it's being calculated from the blurhash, which is *very fast*. Unfortunately, Mastodon doesn't provide blurhashes for avatars so I cannot do that that way. I'll see about quickly grabbing the avg color from the paintable but if it's too time consuming, I'll stay with the accent color

> I was wondering if the numeric inputs are needed?

They were added for accessibility reasons. I can probably do keyboard navigation with the arrow keys but I don't think a screen reader will be able to provide much info from the picker without dedicated rows. Maybe we can go for a more generic approach of dropdowns e.g. "horizontal position: top, center, bottom"... But that also feels clunky

> From that screenshot it looks a bit odd, but hard to judge it without more context. Do you have a link to a post with this? Also, is there any relevant art for this from other clients?

This post: https://mastodon.social/@Mastodon/113272025176822566

![image](https://github.com/user-attachments/assets/92537f39-e92f-4abc-b441-0fc86fb87704)

Will look like this in Tuba:

https://github.com/user-attachments/assets/24dea5c9-654d-41c6-8897-06c48f8fa3da

and clicking it will open the profile.

I think only Phanpy has done it from the clients I can access:

![image](https://github.com/user-attachments/assets/c7f112f1-b52a-4101-b1b5-e0df98263bdb)

The main reason I decided to put it outside the preview card, is the 'horizontal' layout we have when there's no thumbnail, where having an extra row felt a bit off

> It'd be good to have a little bit of padding around the thumbnail, and a slight border radius. I'd also drop the different background color on the textarea.

Border radius might require more thought because it's actually a GtkPicture that takes as much space as it can get currently while keeping the aspect ratio, so I'm not going to block it on this.

I re-added the background color for the sake of indicating the editable space. I'll see about adding a placeholder text instead but it'll be hacky. This is what it looks like without the background:

![Screenshot From 2024-12-03 21-17-42](https://github.com/user-attachments/assets/f24d3190-1368-4477-9f5f-c74aae144367)

> Wrapped

Most interactions was supposed to not be visible if it was empty, which leads me to believe I probably messed up somewhere else.

I worked on this way before it was merged so I didn't have real info. Actually, I still don't have any because my instance hasn't generated them yet. Could you send me yours if you don't mind? Run the following in your browser's console while on mastodon.social:

```js
await fetch("https://mastodon.social/api/v1/annual_reports/", {
    "credentials": "include",
    "headers": {
        "User-Agent": "Mozilla/5.0 (X11; Linux x86_64; rv:134.0) Gecko/20100101 Firefox/134.0",
        "Authorization": `Bearer ${JSON.parse(document.querySelector("#initial-state").textContent).meta.access_token}`,
    },
    "method": "GET",
    "mode": "cors"
}).then((e) => e.json()).then((e) => JSON.stringify(e));
```

Feel free to redact anything you think is private.

Changed it to a 'Style' dropdown with preview:

https://github.com/user-attachments/assets/fa24c2cd-51fc-46af-90d2-1fa81c1c3787

Above is 12px padding

Below is 24px padding

![2024-wrapped](https://github.com/user-attachments/assets/d999f387-4dc2-4873-b2a0-9338c1dafd54)

The export to png happens similarly to taking a screenshot of the widgets, so CSS, fonts, accent colors, dark mode, window size do impact it. Hopefully users realize that the style dropdown is meant for the exports and it's not just a random half-implemented window style chooser. That was mostly the reasoning for the 2 dropdowns prior.

> Are you sure the button needs to say "Schedule and Publish"? My feeling is just "Schedule" would be enough, but maybe it could also say "Publish Later"? As for the title, I think it could just be "Schedule Post", no?

Changed it to "Schedule", though my only concern was that users might not realize that by clicking this button, it will schedule it and close the composer, instead of returning back to the composer and continue editing their post.

> I'd also go for a more standard time chooser (like e.g. the one in Railway), the multi-row layout is a bit hard to parse.

![Screenshot From 2024-12-03 20-03-24](https://github.com/user-attachments/assets/a81c92dc-08c6-418e-bfa8-144d5e9028fe)

Feels a bit empty :thinking:

The timezone dropdown exists mostly for org account where the person managing it wants to set the time in UTC rather than their local timezone which might require them doing timezone calculations themselves.

> The row layout for scheduled posts would look better with a divider line between the header and the post itself. I'd also drop the bold and the colon in the header, and just go for a normal label like "Scheduled for March 20, 2025 ...".

![Screenshot From 2024-12-03 19-50-55](https://github.com/user-attachments/assets/746400da-51fd-44ea-afa8-f63a43a1765d)

LGTM

---

I ended up breaking the feature freeze for 1-2 more things:

- Mutual followings

https://github.com/user-attachments/assets/f0f0f47e-f03f-4bb8-a7df-728bb50527f0

This is similar to twitter's 
![image](https://github.com/user-attachments/assets/c1647be2-eecd-4cc1-b60b-03cc1c44b360)

It shows you which ones of the people you follow also follow this account. I think the way I've done it's a bit too much info but also I don't know where to place it otherwise. Adding it below the 'Follow' button will create spacing between the profile pic and the name/handle. Placing it in the same row as the name/handle (but still positioned below the follow button), will wrap a lot on narrow window sizes + long names. Placing it above/below the fields might be better but also gives it a bit too much attention.

'10 Mutuals' also doesn't feel too descriptive, as it's more like 'people you follow that also follow this person'

- Account suggestions

This is still very WIP so I don't have any screenshots, but mastodon shows a small carousel of suggested accounts (based on the Explore tab's For You accounts) when you don't follow many people:

![image](https://github.com/user-attachments/assets/c70770c8-86aa-473e-ab19-acf3758018ba)

The wip looks like the scheduled posts card with the title "Suggested Accounts", a trash icon button on the end with the tooltip "Don't show me again" and a horizontal scrollable list of accounts similar to the ones in the screenshots above, with left and right osd arrows similar to the gnome-software banners one. It's not a carousel because its carousel page can only have one widget which feels too much for 5+ accounts, 1 per page. While a horizontal scrollable box will show 2-3+ at a time

:shrug: they can both wait for 0.10

**Comment by GeopJr** [comment]
![image](https://github.com/user-attachments/assets/c930ca64-17fc-4256-8dbc-bd68e9eb90e9)

these should probably change positions to: Style ... Save As, Share

**Comment by GeopJr** [comment]
> Account suggestions

Almost done

https://github.com/user-attachments/assets/a50a9ea0-0cba-4ad9-9286-ca842af99a1f

Should probably center the display name
